### ISSUE TITLE
No compatible key exchange algorithms found in /var/www/vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php on line 1589

### ISSUE THREAD

**Issue by appdevcor** [phpseclib/phpseclib#1389]
Hi please some help,

I'm using phpseclib to connect to my sftp using two methods 
1 - using username and password 
2- using private key

both cases i can connect with success using fileZilla but when i trie to connect with php i got this error.
No compatible key exchange algorithms found in /var/www/vendor/phpseclib/phpseclib/phpseclib/Net/SSH2.php on line 1589 and connection failed

this is my code 

<?php

require_once dirname(__FILE__)."/../vendor/autoload.php";
use phpseclib\Crypt\RSA;
use phpseclib\Net\SFTP;

$sftp = new SFTP('sftp.domaine.com');
		$key = new RSA();
		$key->loadKey(file_get_contents(dirname(__FILE__).'/../ptivate.ppk'));

		$logged = $sftp->login('username', $key);
		if (!$logged) {
			echo "not logged \n";
			print_r($sftp->getSFTPErrors());
			
		}

?>

any one had this situation need help please
Thank you

**Comment by appdevcor** [comment]
require_once dirname(__FILE__)."/../vendor/autoload.php";
use phpseclib\Crypt\RSA;
use phpseclib\Net\SFTP;



$sftp = new SFTP('sftp.domaine.com');
		$key = new RSA();
		$key->loadKey(file_get_contents(dirname(__FILE__).'/../ptivate.ppk'));

		$logged = $sftp->login('username', $key);
		if (!$logged) {
			echo "not logged \n";
			print_r($sftp->getSFTPErrors());
			
		}

**Comment by terrafrost** [comment]
What would be most relevant are the SSH logs. eg. do `define('NET_SSH2_LOGGING', 2)` at the top and then `echo $sftp->getLog()` after the failed login. Then post the results on pastebin.com and post the link here.

**Comment by appdevcor** [comment]
this is the link : https://pastebin.com/rn2dAZLg
Thank you so much for help

**Comment by terrafrost** [comment]
The server only supports the following key exchange algorithms:

- ecdh-sha2-nistp256
- ecdh-sha2-nistp384
- ecdh-sha2-nistp521.

Currently no version of phpseclib supports those. Not the 1.0 branch, the 2.0 branch or the master branch.

There is an experimental branch at https://github.com/terrafrost/phpseclib/tree/diffie-hellman that does add support for those key exchange methods but currently the unit tests for that branch are failing. At least they fail when GNU Parallels is used on Travis CI. When it isn't used they're passing.

I do intend to merge this branch into the master branch but I need to figure out what's up with the unit tests and then add new unit tests for the new features that that branch adds.

**Comment by appdevcor** [comment]
Thank you again for your help,

can i get the list of key exchange algorithms supported by phpseclib ?
and do you khnow an alternative way to connect using php with this type of sftp 
thank you for helping me

**Comment by terrafrost** [comment]
> can i get the list of key exchange algorithms supported by phpseclib ?

- diffie-hellman-group1-sha1
- diffie-hellman-group14-sha1
- diffie-hellman-group-exchange-sha1
- diffie-hellman-group-exchange-sha256

curve25519-sha256@libssh.org is supported, as well, if you're using libsodium with the _old_ OOP API (ie. not the libsodium that PHP 7.2 includes by default). This needs to be updated.

> do you khnow an alternative way to connect using php with this type of sftp

tbh I'd recommend the experimental branch.

There are three main ways to do SSH in PHP

1. phpseclib
2. libssh2. Which supports even less key exchange algorithms than the currently released versions of phpseclib do. From https://www.php.net/manual/en/function.ssh2-connect.php it supports the following:
    - diffie-hellman-group1-sha1
    - diffie-hellman-group14-sha1
    - diffie-hellman-group-exchange-sha1
  So whereas the current versions of phpseclib support four different key exchange algorithms libssh2 only supports three. And the three that it supports are also supported by phpseclib. The experimental branch, in contrast, supports fourteen.
3. OpenSSH via the CLI

**Comment by appdevcor** [comment]
thank you for all this détails one more request please, How can i install the expérimental branch via composer.

**Comment by terrafrost** [comment]
```
{
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/terrafrost/phpseclib"
        }
    ],
    "require": {
        "phpseclib/phpseclib": "dev-diffie-hellman"
    }
}
```

**Comment by appdevcor** [comment]
Thank you terrafrost for your help, You saved me.

**Comment by gmariani** [comment]
I'd like to chime in on this as I'm running into this same issue as well. I did learn that phpseclib only supports the old version of libsodium. I was able to update SSH2.php without too much hassle to get it to work with the newer version of libsodium. This solved my KEX issue, but then I ran into a MAC issue.

**phpseclib MAC Support**: hmac-sha2-256, hmac-sha1-96, hmac-sha1, hmac-md5-96, hmac-md5
**My server MAC Support**: hmac-sha2-512-etm@openssh.com, hmac-sha2-256-etm@openssh.com, umac-128-etm@openssh.com

As you can see there isn't a matching MAC. But unfortunately I can't enable the required MAC as they are insecure. Is there a possibility of updating the MAC list? The MAC list I'm going off of is based on https://github.com/arthepsy/ssh-audit .

**SSH Audit Error messages for currently supported MACs**:
- hmac-sha2-256: WARN_ENCRYPT_AND_MAC
- hmac-sha1-96: FAIL_OPENSSH67_UNSAFE, WARN_OPENSSH72_LEGACY, WARN_ENCRYPT_AND_MAC, WARN_HASH_WEAK
- hmac-sha1: WARN_ENCRYPT_AND_MAC, WARN_HASH_WEAK
- hmac-md5-96: FAIL_OPENSSH67_UNSAFE, WARN_OPENSSH72_LEGACY, WARN_ENCRYPT_AND_MAC, WARN_HASH_WEAK
- hmac-md5: FAIL_OPENSSH67_UNSAFE, WARN_OPENSSH72_LEGACY, WARN_ENCRYPT_AND_MAC, WARN_HASH_WEAK

**Legend**
FAIL_OPENSSH67_UNSAFE = 'removed (in server) since OpenSSH 6.7, unsafe algorithm'
WARN_OPENSSH72_LEGACY = 'disabled (in client) since OpenSSH 7.2, legacy algorithm'
WARN_ENCRYPT_AND_MAC  = 'using encrypt-and-MAC mode'
WARN_HASH_WEAK        = 'using weak hashing algorithm'

I realize the complexity in adding a MAC (i have no idea where to begin there), but if it is at all possible, I'd like to request it. If I can help, i can try there as well too. Thanks!

**Comment by terrafrost** [comment]
I figure I'll update the MACs after the I submit the PR for the key exchange stuff. I'm pretty close to being done with that. I've added new unit tests for the key exchange changes and it looks like two of the unit tests are failing so I just gotta get them passing and then I'll submit a PR.

I suppose I could skip the PR and just commit directly to master but for bigger changes I like the PR process still. Among other things it's an opportunity for me to provide some limited documentation on the new features being added and it's a sorta announcement of sorts, as well.

Once the PR is submitted I'll probably let it sit for a week or so so people can chime in on it and then I'll merge it. I figure I'll work on the HMAC stuff while it's sitting around, pending merging.

I have a few other smaller changes I want to do (add Curve448, maybe make the libsodium key format "hidden" since I have a strong suspicion most any random 32-byte string would be identified as one, etc) but after those I figure I'll create a new 3.0 branch. I'll let it sit for a while as I work on the documentation and then I'll release 3.0.0.

I don't want to create a 3.0 branch quite yet because once I create that 3.0 branch I'm basically committing to that API in my mind. The master branch has an unstable API that's subject to change - the 3.0 branch won't.

TLDR the MAC algorithms are slated to receive an overhaul and 3.0.0 will probably be released in a few months time.

**Comment by terrafrost** [comment]
Also, @gmariani - does your server support chacha20-poly1305 or aes128/256-gcm? If so then the negotiated MAC algorithms aren't utilized - as those are AEAD algorithms they kinda eliminate the need for a MAC.

Obviously that doesn't eliminate the need for updated MAC algorithms in the general case but it might address your specific need for them? idk lol

**Comment by bantu** [comment]
@gmariani Can you share your patch for updating SSH2 for new versions of libsodium?
Or is that something already accounted for in the master branch, @terrafrost ?

**Comment by terrafrost** [comment]
@bantu - the master branch (well diffie-hellman branch, pending merging) eliminates the need for such a patch but one is still needed for the 2.0 branch. I'll probably do the 2.0 branch at some point.

**Comment by bantu** [comment]
@terrafrost Is it only this one-line change? https://github.com/phpseclib/phpseclib/pull/1390/commits/66efabda53a68bc6f87abea4c1d585e0398f8570

**Comment by terrafrost** [comment]
@bantu - no - it's a few more lines than that but the changes are of comparable complexity lol.

eg. lines like `$eBytes = \Sodium\crypto_box_publickey_from_secretkey($x);` would need to be updated as well in the 2.0 branch.

In the master branch, those lines aren't even present, anymore, because they've already been replaced with the new DH class.

**Comment by bantu** [comment]
@terrafrost Okay, I see. Thanks.

**Comment by Echron** [comment]
@terrafrost What is the status of this?

**Comment by terrafrost** [comment]
@Echron - the 3.0 branch should solve most of these issues. The biggest culprit of this issue that that branch currently leaves unresolved are servers that _require_ compression.

The 3.0 branch is ready to be used but it has not had a tagged release yet. I'm holding off on tagging a release until I've completed the documentation for the 3.0 branch. In the mean time you should be able to include it by requiring `dev-3.0`.

**Comment by IceLitFire** [comment]
@terrafrost when will phpseclib 3.0 be released?  Will it cover these kex?

ECDH-SHA2-NISTP256
ECDH-SHA2-NISTP384
ECDH-SHA2-NISTP512

We're stuck with a partner who is requiring these suddenly and our phpseclib is not capable.

**Comment by terrafrost** [comment]
3.0.0 was released on Dec 17. The current version is now 3.0.2. The documentation for 3.0 lives at https://phpseclib.com/

If you want to use the 2.0 API but use the 3.0 features consider using https://github.com/phpseclib/phpseclib2_compat (which has phpseclib 3.0 as a dependency)

**Comment by terrafrost** [comment]
@IceLitFire yes - 3.0 supports CDH-SHA2-NISTP256, ECDH-SHA2-NISTP384 and ECDH-SHA2-NISTP512
